Mitigation for JWT Format Confusion #2 (#11)
yaronf/draft-sheffer-oauth-rfc8725bis: draft-sheffer-oauth-rfc8725bis.md

For mitigations, see {{algorithm-verification}}.

## JWT Format Confusion

> Some implementations support both Compact and JSON Serialization formats. If an application verifies a JWT in JSON format but extracts claims by parsing it as a Compact JWT (e.g., via string splitting), an attacker can craft a valid JSON JWS with a forged payload. This mismatch in format handling can lead to authentication bypass or impersonation.

Author: I thought JSON serialization was a valid format. No?

> Some implementations support both Compact and JSON Serialization formats. If an application verifies a JWT in JSON format but extracts claims by parsing it as a Compact JWT (e.g., via string splitting), an attacker can craft a valid JSON JWS with a forged payload. This mismatch in format handling can lead to authentication bypass or impersonation.

Reviewer: Quoting RFC 7519: "JWTs are always represented using the JWS Compact Serialization or the JWE Compact Serialization."

> Some implementations support both Compact and JSON Serialization formats. If an application verifies a JWT in JSON format but extracts claims by parsing it as a Compact JWT (e.g., via string splitting), an attacker can craft a valid JSON JWS with a forged payload. This mismatch in format handling can lead to authentication bypass or impersonation.

Author: While 7519 defines a JWT and that it is the compact serialization, 7515 (JWS) and 7516 (JWE) both define a compact and JSON serialization, and the attack was a library that would take both of those serializations, which sortof makes sense as they are both defined in 7515 and 7516.

https://datatracker.ietf.org/doc/html/rfc7515#section-7

https://datatracker.ietf.org/doc/html/rfc7516#section-7

@selfissued what are your thoughts?

> Some implementations support both Compact and JSON Serialization formats. If an application verifies a JWT in JSON format but extracts claims by parsing it as a Compact JWT (e.g., via string splitting), an attacker can craft a valid JSON JWS with a forged payload. This mismatch in format handling can lead to authentication bypass or impersonation.

Reviewer 2: ```suggestion
Some JWS implementations support both the Compact and JSON Serializations. While JWTs MUST use the Compact Serialization, if an application by mistake verifies a JWT using the JSON Serialization but extracts claims by parsing it as a JWT using the Compact Serialization (e.g., via string splitting), an attacker can craft a valid JSON JWS with a forged payload. This mismatch in format handling can lead to authentication bypass or impersonation.
```

> Some implementations support both Compact and JSON Serialization formats. If an application verifies a JWT in JSON format but extracts claims by parsing it as a Compact JWT (e.g., via string splitting), an attacker can craft a valid JSON JWS with a forged payload. This mismatch in format handling can lead to authentication bypass or impersonation.

Reviewer 2: My thoughts are in my change suggestion about JWTs containing only letters, numbers, dash, underscore, and period.

Thus, rejecting inputs with a `p2c` (PBES2 Count) value over 1,200,000 (double that) is RECOMMENDED.

## Check JWT Format Type {#token-format}

> Implementations are MUST to confirm if the JWT is a supported format before parsing the JWT. If the implementation only supports the compact format, the token will start with 'e' (ASCII 0x65) and if the implementation supports the JSON format, the token will start with '{' (ASCII 0x7B).

Reviewer: ```suggestion
Implementations MUST confirm that the JWT is a supported format before parsing the JWT. If the implementation only supports the compact format, the token will start with 'e' (ASCII 0x65) and if the implementation supports the JSON format, the token will start with '{' (ASCII 0x7B).
```

> Implementations are MUST to confirm if the JWT is a supported format before parsing the JWT. If the implementation only supports the compact format, the token will start with 'e' (ASCII 0x65) and if the implementation supports the JSON format, the token will start with '{' (ASCII 0x7B).

Author: That is a better check, thanks Mike.

> Implementations are RECOMMENDED to confirm if the JWT is a supported format before parsing the JWT. If the implementation only supports the compact format, the token with start with e (ASCII 0x65) and if the implementation supports the JSON format, the token will start with { (ASCII 0x7B). 

Reviewer: RECOMMENDED means SHOULD. Why not make it a MUST?

> Implementations are RECOMMENDED to confirm if the JWT is a supported format before parsing the JWT. If the implementation only supports the compact format, the token with start with e (ASCII 0x65) and if the implementation supports the JSON format, the token will start with { (ASCII 0x7B). 

Reviewer: ```suggestion
Implementations are RECOMMENDED to confirm if the JWT is a supported format before parsing the JWT. If the implementation only supports the compact format, the token will start with 'e' (ASCII 0x65) and if the implementation supports the JSON format, the token will start with '{' (ASCII 0x7B). 
```

{{OWASP-Password-Storage}} states that an iteration count of 600,000 is required when using HMAC-SHA-256 to achieve FIPS-140 compliance.
Thus, rejecting inputs with a `p2c` (PBES2 Count) value over 1,200,000 (double that) is RECOMMENDED.

## Check JWT Format Type {#token-format}

> Implementations MUST confirm if the JWT is a supported format before parsing the JWT. If the implementation only supports the compact format, the token will start with 'e' (ASCII 0x65) and if the implementation supports the JSON format, the token will start with '{' (ASCII 0x7B).

Reviewer 2: ```suggestion
Implementations MUST confirm the JWT is in a legal format while parsing it. Legal JWTs contain only the ASCII characters for letters, numbers, dash, underscore, and period.  Content with any other characters - especially braces and quotation marks - is not a JWT and MUST be rejected.
```

Because of leading spaces being legal in JSON, neither of the "e" nor "{" statements is true.
